Userland/Libraries/LibGfx/ImageFormats/FLICLoader.cpp

```diff
@@ -0,0 +1,388 @@
+                    u8 g = TRY(stream->read_value<u8>()) << 2;
+                    u8 b = TRY(stream->read_value<u8>()) << 2;
+                    current_frame->color_map[color_index++] = { r, g, b };
+                }
+            }
+            break;
+        }
+
+        case ChunkType::FLI_COPY: {
+            dbgln_if(FLIC_DEBUG, "Decoding copy_buffer {}x{} buffer_size={}", context.width, context.height, stream->size());
+            context.frame_buffer = TRY(Bitmap::create(BitmapFormat::BGRA8888, { context.width, context.height }));
+            for (size_t y = 0; y < context.height; ++y) {
+                for (size_t x = 0; x < context.width; ++x) {
+                    context.frame_buffer->set_pixel(x, y, current_frame->color_map[TRY(stream->read_value<u8>())]);
+                }
+            }
+            break;
+        }
+
+        case ChunkType::DELTA_FLI: {
+            dbgln_if(FLIC_DEBUG, "Decoding delta_fli {}x{} buffer_size={}", context.width, context.height, stream->size());
+            auto lines_to_skip = TRY(stream->read_value<LittleEndian<u16>>());
+            auto number_of_lines = TRY(stream->read_value<LittleEndian<u16>>());
+            auto& previous_frame = context.frames.at(frame_index - 1);
+
+            for (size_t line = lines_to_skip; line < lines_to_skip + number_of_lines; line++) {
+                size_t x = 0;
+                auto packets = TRY(stream->read_value<u8>());
+
+                for (u8 packet = 0; packet < packets; packet++) {
+                    auto col_skip_count = TRY(stream->read_value<u8>());
+                    auto type = TRY(stream->read_value<i8>());
+                    x += col_skip_count;
+
+                    // copy next pixels
+                    if (type > 0) {
+                        auto dest_x = x;
+                        for (i8 n = 0; n < type; n++) {
+                            auto color = TRY(stream->read_value<u8>());
+                            context.frame_buffer->set_pixel(dest_x++, line, previous_frame->color_map[color]);
+                        }
+                    } else if (type < 0) {
+                        type = -type;
+                        // repeat next pixels
+                        auto color = TRY(stream->read_value<u8>());
+                        auto dest_x = x;
+                        for (i8 n = 0; n < type; n++) {
+                            context.frame_buffer->set_pixel(dest_x++, line, previous_frame->color_map[color]);
+                        }
+                    }
+
+                    x += type;
+                }
+            }
+            break;
+        }
+
+        default: {
+            dbgln_if(FLIC_DEBUG, "Unknown main chunk {:#x}", chunk->type);
+            break;
+        }
+        }
+    }
+
+    context.state = FLICLoadingContext::State::FrameComplete;
+
+    return {};
+}
+
+static ErrorOr<void> read_subchunk(FLICLoadingContext& context)
+{
+    auto& current_frame = context.frames.last();
```

Comment:
I'm realizing it right now, but why are frames and chunks heap-allocated?
I don't think this is necessary